Loader and reduction for Lagrange at ILL

MathieuTillet opened the pull request:
**Description of work.**
This PR introduces a loader and a reduction algorithm for Lagrange data.
Lagrange data being a instrument scan stored in an ASCII file, it was considered easier to manage the loading side directly in the Python algorithm. Due to this, it is not possible to load the data through a general loader (but loading non-reduced data is not very informative, and could be done by running the reduction algorithm without any reduction step).

**To test:**


Run ILLLagrange with

    SampleRuns: 012882:012884,
    ContainerRuns: 012869:012871,
    CorrectionFile: correction-water-cu220-2020.txt

and check the 1D plot make  sense. Also check that an invisible workspace group with intermediate reduction states has been created.

Fixes #34746 .

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

MathieuTillet committed:
Create algorithm for Lagrange reduction

MathieuTillet committed:
Interface Lagrange with Drill

MathieuTillet committed:
Better merging process for adjacent points

MathieuTillet committed:
Offset axis by incident energy

MathieuTillet committed:
Save intermediate results as invisible ws

MathieuTillet committed:
Reduce data using mantid

Rather than relying on numpry everywhere

MathieuTillet committed:
Determine header size and columns position in file

Doing so automatically makes the loading more robust in case of changing
in the structure of the ASCII files

MathieuTillet committed:
Set constants as class variables

MathieuTillet committed:
Set constants as class variables in Lagrange

And clarify some functions name and doc

MathieuTillet committed:
Tests for ILLLagrange

MathieuTillet committed:
[LAGRANGE] Add option to see incident energy

MathieuTillet committed:
[LAGRANGE] Correctly set energy units

MathieuTillet committed:
[LAGRANGE] Add an option to convert energy to cm-1

MathieuTillet committed:
[LAGRANGE] Add tests

MathieuTillet committed:
[LAGRANGE] Add test data

MathieuTillet committed:
[LAGRANGE] Clarify incident energy option

MathieuTillet committed:
[LAGRANGE] Add release note

MathieuTillet committed:
[LAGRANGE] Rename algorithm file to correct name

MathieuTillet committed:
[LAGRANGE] Add algorithm documentation

darominski commented on docs/source/release/v6.6.0/Direct_Geometry/General/New_features/34748.rst:
```diff
@@ -0,0 +1 @@
+Added `ILLLagrange`, which reduces data from IN1 - Lagrange at ILL.
```
Could you add here a reference to the algorithm so that it connects to the documentation?

darominski commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ILLLagrange.py:
```diff
@@ -0,0 +1,308 @@
+
+        # empty cell treatment, if there is any
+        if empty_cell_files[0]:
+            # load and format empty cell
+            empty_cell_data = self.load_and_concatenate(empty_cell_files)
+            empty_cell_data = self.merge_adjacent_points(empty_cell_data)
+            energy, detector_counts, errors = self.get_counts_and_errors(empty_cell_data)
+
+            self.empty_cell_ws = "__" + self.output_ws_name + "_rawEC"
+
+            CreateWorkspace(outputWorkspace=self.empty_cell_ws,
+                            DataX=energy,
+                            DataY=detector_counts,
+                            DataE=errors,
+                            UnitX="Energy")
+
+            self.intermediate_workspaces.append(self.empty_cell_ws)
+
+            # correct by water
+            if self.water_correction is not None:
+                correctedECname = "__" + self.output_ws_name + "_waterCorrectedEC"
+                self.correct_data(self.empty_cell_ws, correctedECname)
+
+                self.intermediate_workspaces.append(correctedECname)
```
Do you think the empty cell treatment could be extracted to a separate method to keep the PyExec as clean as possible?

darominski commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ILLLagrange.py:
```diff
@@ -0,0 +1,308 @@
+
+        return data[:current_writing_index + 1]
+
+    def load_correction(self, file):
```
Since this wrapper is so thin, is it even needed here?

darominski commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ILLLagrange.py:
```diff
@@ -0,0 +1,308 @@
+            if point[0] - data[current_writing_index][0] < self.EPSILON:
+                data[current_writing_index][1] = point[1]
+                data[current_writing_index][2] = point[2]
+                # TODO decide on a merging behaviour for multiple points
```
The behaviour here does not seem to correspond to the comment below, as the data is currently overwritten (points being removed), if I understand this correctly. There is no calculation of the mean at the moment. I presume also this is the part that awaits user's response. For now, could you correct the comment here so it better corresponds to the current behaviour?

darominski commented on docs/source/algorithms/ILLLagrange-v1.rst:
```diff
@@ -0,0 +1,94 @@
+-----------
+This algorithm reduces data from IN1 - Lagrange.
+
+It can take as input multiples files from a single monochromator scan, the associated empty cell runs, a correction file
```
typo: multiple
after last comma: **and** a correction file

darominski commented on docs/source/algorithms/ILLLagrange-v1.rst:
```diff
@@ -0,0 +1,94 @@
+This algorithm reduces data from IN1 - Lagrange.
+
+It can take as input multiples files from a single monochromator scan, the associated empty cell runs, a correction file
+if needed. Reduction consists in multiplying the data by the correction factor from the correction file, and subtracting
```
In this context, consisting **of** might fit better rather than **in**.

darominski commented on docs/source/algorithms/ILLLagrange-v1.rst:
```diff
@@ -0,0 +1,94 @@
+The X-axis of the result data can be offset by the incident energy using `UseIncidentEnergy`, and be converted to wave
+number instead of the default energy using `ConvertToWaveNumber`.
+
+All the files in each field are merged together in a single curve, with very close points being merged together to avoid
```
The current behaviour removes these potentially problematic points rather than merging them, which behaviour would be more desired (probably up to the user)?

darominski commented on docs/source/algorithms/ILLLagrange-v1.rst:
```diff
@@ -0,0 +1,94 @@
+Usage
+-----
+
+**Simple Example**
```
Since these test codes do not check for any particular output, do you think it makes sense to make them a code block instead?

darominski reviewed (changes requested):
Thank you for the implementation of the Lagrange workflow. I have a couple pedantic comments dispersed through the documentation and the code. The major point I would like to clarify is the merging of problematic points.

Minor general comments:
1. Could you add type hints to the methods?
2. Do you think it makes sense tot provide a workflow diagram? The reduction is quite straightforward and linear, so it is just a question to consider.

darominski commented:
Since we didn't merge it before @MathieuTillet left, I am going to fix the issues I found. We also will not wait longer for the user's response and will handle them later, when comments come in.

darominski committed:
[LAGRANGE] Moved empty cell processing to own method

darominski committed:
[LAGRANGE] Removed thin wrapper for loading water correction

darominski committed:
[LAGRANGE] Updated merging behaviour comment

darominski committed:
[LAGRANGE] Doc typo fixes, suppression of testcode

in favour of using code-block since the code was not being tested except
for raising exceptions.

darominski committed:
[LAGRANGE] Added link to algorithm in the release note

darominski reviewed (commented):
Taken care of my own comments.

gui-co commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ILLLagrange.py:
```diff
@@ -0,0 +1,308 @@
+            energy[index] = line[0] - offset
+
+            detector_counts[index] = line[2] / line[1]
+            errors[index] = np.sqrt(line[2]) / line[1]
```
After discussion with the scientists, we should calculate the error on the monitor counts as well and propagate it.

gui-co commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ILLLagrange.py:
```diff
@@ -0,0 +1,308 @@
+                monitor_counts_col = self.MONITOR_COUNT_COL_DEFAULT
+
+            # load the relevant data, which position we just determined - angle, monitor count, detector count
+            data = np.loadtxt(file,
+                              dtype=float,
+                              skiprows=header_index + 2,
+                              usecols=(energy_col, monitor_counts_col, detector_counts_col))
+
+            loaded_data = np.vstack((loaded_data, data)) if loaded_data is not None else data
+
+        return loaded_data
```
Here, if one of the file cannot be read, this will be silently ignored. If all files cannot be read, an obscure error will come from the workspace creation line 100.

I would add a check on `data` to verify that something was actually read from the file and a final check on `loaded_data` as well.

gui-co commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ILLLagrange.py:
```diff
@@ -0,0 +1,308 @@
+
+        # load correction
+        if correction_file:
+            self.water_correction = np.loadtxt(correction_file)
```
I would also add a try/catch and a check on `self.water_correction` to properly report any failure when reading the file.

gui-co commented on docs/source/algorithms/ILLLagrange-v1.rst:
```diff
@@ -0,0 +1,93 @@
+-----------
+This algorithm reduces data from IN1 - Lagrange.
+
+It can take as input multiples files from a single monochromator scan, the associated empty cell runs, and a correction file
```
```suggestion
It can take as input multiple files from a single monochromator scan, the associated empty cell runs, and a correction file
```

gui-co reviewed (commented):
I added few comments in the code. My main concern is with error handling when reading the ascii files.

darominski committed:
[LAGRANGE] Corrected typo in doc

Co-authored-by: Guillaume Communie <35613913+gui-co@users.noreply.github.com>

darominski committed:
[LAGRANGE] Updated copyright date

darominski committed:
[LAGRANGE] Added try-except for loading water correction data

which necessitated extracting loading into own method (again) to keep PyExec clean.

darominski committed:
[LAGRANGE] Added type hints

darominski committed:
[LAGRANGE] Changed algorithm name for consistency

darominski committed:
[LAGRANGE] Include monitor errors in the detector error propagation

gui-co reviewed (approved):
Thanks for the changes.
